### PR TITLE
[MRG] feat (bandits) Randomized agents

### PR DESCRIPTION
**Pull request by sauxpa**
* Add RandomizedAgent class, inspired by IndexAgent, with default policy being EXP3 with standard, anytime hyperparam,
* Add AdversarialBandit environment, where all potential rewards are fixed in advance (oblivious opponent). 
* Refactor envs/bandits/bandits.py to remove redundant syntax when instantiating a child class.
* Refactor examples/demo_bandits to display pseudo-regret (best mean - mean of chosen arm) rather than regret (best mean - reward of chosen arm).
   * Pseudo-regret is the standard measure in bandit, and it is by definition nondecreasing, whereas regret as it is now is noisy.
* Update tests
   * Change utils/check_bandit_agent to test on BernoulliBandit rather than NormalBandit. 
      * Reason is some algo (EXP3) crucially rely on bounded rewards (by default [0, 1]). I think it's safer to test by default on Bernoulli rather than Gaussian. Also I explicitly pass the env in argument in the test for readability. 
   * Add test for BernoulliBandit env. 
   * Explicitly fixes the seed (add seed argument).

### CODE REVIEW

**Comment by sauxpa** [comment]
Windows build now fails but seems to be unrelated to my commit (used to pass on previous commits, now fails after reverting).

File: rlberry/manager/evaluation.py [hunk 1]
```diff
@@ -304,6 +304,9 @@ def plot_writer_data(
     if ax is None:
         figure, ax = plt.subplots(1, 1)
 
+    # PS: in the next release of seaborn, ci should be deprecated and replaced
+    # with errorbar, which allows to specifies other types of confidence bars,
+    # in particular quantiles.
```
Comment (TimotheeMathieu) [inline rlberry/manager/evaluation.py:309]: Nice ! we may want to change to quantile then. As of now, we chose the 'sd' ci because choosing the permutation ci (ie argument ci = 75 for instance) is costly with large dataset.

File: rlberry/utils/check_bandit_agent.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
-from rlberry.envs.bandits import NormalBandit
+from rlberry.envs.bandits import BernoulliBandit
 from rlberry.manager import AgentManager
 
 
-def check_bandit_agent(Agent, environment=NormalBandit):
+def check_bandit_agent(Agent, environment=BernoulliBandit, seed=None):
```
Comment (TimotheeMathieu) [inline rlberry/utils/check_bandit_agent.py:5]: the seed has to be fixed. It can be any int, but it can't be None because if it is None, there would be reproducibility issues. Maybe you can change the default to some int (42 is a good choice :) )

File: rlberry/envs/tests/test_bandits.py [hunk 3]
```diff
@@ -1,20 +1,35 @@
...
     assert np.abs(np.mean(sample) - 1) < 0.1
 
 
 def test_cor_normal():
     env = CorruptedNormalBandit(means=[0, 1], cor_prop=0.1)
-    safe_reseed(env, Seeder(42))
+    safe_reseed(env, Seeder(TEST_SEED))
 
     sample = [env.step(1)[1] for f in range(1000)]
     assert np.abs(np.median(sample) - 1) < 0.5
```
Comment (TimotheeMathieu) [inline rlberry/envs/tests/test_bandits.py:35]: Maybe we can also include some generic bandit testing using something like in PR #110 ? You can do it here or I will do a separate PR.

File: rlberry/agents/bandits/index_agents.py [hunk 4]
```diff
@@ -34,9 +34,8 @@ class IndexAgent(BanditWithSimplePolicy):
 
     name = "IndexAgent"
 
-    def __init__(self, env, index_function=lambda rew, t: np.mean(rew), **kwargs):
+    def __init__(self, env, index_function=None, **kwargs):
         BanditWithSimplePolicy.__init__(self, env, **kwargs)
-        self.n_arms = self.env.action_space.n
```
Comment (TimotheeMathieu) [inline rlberry/agents/bandits/index_agents.py:39]: If you rename all the "index" to "indices", shouldn't it be "indices_function" and "Indice_Agent" ?

File: rlberry/agents/bandits/randomized_agents.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        See :class:`~rlberry.envs.bandits.Bandit`.
+
+    index_function : callable or None, default = None
+        Compute the index for an arm using the past rewards and sampling
+        probability on this arm and the current time t.
+        If None, use loss-based importance weighted estimator.
+
+    prob_function : callable or None, default = None
+        Compute the sampling probability for an arm using its index.
+        If None, use exponential softmax probabilities with any learning rate.
```
Comment (TimotheeMathieu) [inline rlberry/agents/bandits/randomized_agents.py:24]: I don't understand the bit about "with any learning rate", can you explain more. Maybe you can say "with learning rate from ..." and cite an article.

**Review by TimotheeMathieu — COMMENTED** [review]
Thank you for this and thank you for the changes to the bandit tests. 
I have some comments and otherwise I also have a question. If I understand right, Exp3 is about adversarial bandits. Would it be possible to have some sort of "adversarial environment" and do an example on it ? That would be nice and that would explain why Exp3. I think one important thing for bandits in rlberry is also to provide some benchmark environments for difficult cases (adversarial, outliers, privacy...) on which the community is currently doing some research.

File: rlberry/utils/check_bandit_agent.py (hunk 2 above)
Comment (sauxpa) [reply rlberry/utils/check_bandit_agent.py]: Agreed, done :)

File: rlberry/envs/tests/test_bandits.py (hunk 3 above)
Comment (sauxpa) [reply rlberry/envs/tests/test_bandits.py:35]: Agreed, I'll have a look. Maybe a separate PR for clarity? Ideally it would be great to agree on a good set of tests for bandits and just copy-paste those when we implement more agents, so I think it is worth dedicating some time on this

File: rlberry/agents/bandits/randomized_agents.py (hunk 5 above)
Comment (sauxpa) [reply rlberry/agents/bandits/randomized_agents.py:24]: It's just bad writing from me. I'll resubmit (need to make a small change to the learning rate definition actually -- there are quite a few in the literature, I'm trying to pick a somewhat standard one).

File: rlberry/manager/evaluation.py (hunk 1 above)
Comment (sauxpa) [reply rlberry/manager/evaluation.py:309]: Absolutely, when it's out I'd much rather see quantiles (the update is already on github btw, just not released yet to the latest version of seaborn).

File: rlberry/envs/tests/test_bandits.py (hunk 3 above)
Comment (TimotheeMathieu) [reply rlberry/envs/tests/test_bandits.py:35]: Ok. Ideally, the generic check_bandit function should test only basics properties (reproducibility, save & load and things like that) but maybe we want to keeep some specific tests for each bandit, for instance test that it works well on subgaussian and not only on bernoulli, test that adversarial bandits work on adversarial problems...

File: rlberry/agents/bandits/index_agents.py (hunk 4 above)
Comment (TimotheeMathieu) [reply rlberry/agents/bandits/index_agents.py:39]: Ok, maybe I should read the Oxford dictionary some more ;)

**Comment by sauxpa** [comment]
@TimotheeMathieu : I have given some thoughts on implementing an "adversarial" environment, below is a short recap for both of us.

* The "true" adversarial environment would be one where the user is asked to manually provide the rewards, either from the start (oblivious opponent) or actively during the run of the algorithm (adaptive opponent). 
* Many paper in this literature run experiments on non-stationary rather truly adversarial environments, where typically one arm is the best for many rounds and then becomes the worst (see e.g the experiments in the Tsallis-INF paper).
* Another related problem is that of minimax lower bounds: the supremum of regret at horizon T over a certain family of bandit instances is attained for a certain value of the gap, typically O(\sqrt{1/T}). We might want to have some experiments on it also, but I'll skip it for now.

What I plan is to add an oblivious opponent environment. The adaptive one is more cumbersome and not really useful IMO.
That would give us already much flexibility to design interesting examples. Adding non-stationary bandits in rlberry is a whole topic in itself, and eventually I think we will want to have dedicated environments for them too, but in the meantime we should be able to reproduce standard experiments in the adversarial literature.

**Review by TimotheeMathieu — APPROVED** [review]
LGTM, thanks.
